### PR TITLE
Generalized line unfolding in parser (issue #8)

### PR DESCRIPTION
**Pull request by gracinet**
Ignoring (CR)LF followed by a single whitespace char is now done in peek()
and therefore is enforced in all processing.

This version has been able to parse a full ics file exported from Google (the very one with which I encountered issue #8).
This is work in progress, and my actual first non utterly trivial code in Rust, don't hesitate to say if the approach taken doesn't suit you, or if assumptions about efficiency are wrong etc.

Some choices:

- in `consume_while()`, tried to minimize the copying operations, but it can't return a str (slice) 
  anymore in any case.
- for simplicity, I discarded CR completely from peek_at(). If you agree with this, some conditions 
  downstream (consume_eol()...) can be simplified.

### CODE REVIEW

File: src/vobject/lib.rs [hunk 1]
```diff
@@ -149,8 +178,8 @@ impl<'s> Parser<'s> {
     }
 
     fn consume_char(&mut self) -> Option<char> {
-        match self.peek() {
-            Some(x) => { self.pos += x.len_utf8(); Some(x) },
+        match self.peek_at(0) {
```
Comment (untitaker) [inline src/vobject/lib.rs]: Unnecessary change?

File: src/vobject/lib.rs [hunk 2]
```diff
@@ -269,20 +317,20 @@ impl<'s> Parser<'s> {
             x > '\u{1F}'
         };
 
-        if self.peek() == Some('"') {
+        if self.peek() == Some(('"', 1)) {
```
Comment (untitaker) [inline src/vobject/lib.rs]: Do we actually require offset == 1 in this conditional?

File: src/vobject/lib.rs [hunk 3]
```diff
@@ -269,20 +317,20 @@ impl<'s> Parser<'s> {
...
         } else {
-            Ok(self.consume_while(|x| qsafe(x) && x != ';' && x != ':').to_owned())
+            Ok(self.consume_while(|x| qsafe(x) && x != ';' && x != ':'))
         }
     }
 
     fn consume_param<'a>(&'a mut self) -> ParseResult<(String, String)> {
         let name = try!(self.consume_param_name());
-        let value = if self.peek() == Some('=') {
+        let value = if self.peek() == Some(('=', 1)) {
```
Comment (untitaker) [inline src/vobject/lib.rs]: Do we actually require offset == 1 in this conditional?

File: src/vobject/lib.rs [hunk 4]
```diff
@@ -298,7 +346,7 @@ impl<'s> Parser<'s> {
 
     fn consume_params(&mut self) -> HashMap<String, String> {
         let mut rv: HashMap<String, String> = HashMap::new();
-        while self.peek() == Some(';') {
+        while self.peek() == Some((';', 1)) {
```
Comment (untitaker) [inline src/vobject/lib.rs]: Do we actually require offset == 1 in this conditional?

**Comment by untitaker** [comment]
Thanks, this is fantastic! I haven't looked closely through consume_while yet but I wouldn't have cared about the inefficient method.

Note this was one of the first things I did in Rust (and I come from a higher-level language) so code might actually be unidiomatic and imperformant (because I didn't care).

File: src/vobject/lib.rs [hunk 5]
```diff
@@ -185,12 +214,39 @@ impl<'s> Parser<'s> {
...
-        while !self.eof() && test(self.peek().unwrap()) {
-            self.consume_char();
+    // GR this used to return just a slice from input, but line unfolding
+    // makes it contradictory, unless one'd want to rescan everything.
+    // Since actually useful calls used to_owned() on the result, which
+    // does copy into a String's buffer, let's create a String right away
+    // implementation detail : instead of pushing char after char, we
+    // do it by the biggest contiguous slices possible, because I believe it
+    // to be more efficient (less checks for reallocation etc).
+    fn consume_while<'a, F: Fn(char) -> bool>(&'a mut self, test: F) -> String {
```
Comment (untitaker) [inline src/vobject/lib.rs]: The lifetime here is now useless.

File: src/vobject/lib.rs [hunk 6]
```diff
@@ -185,12 +214,39 @@ impl<'s> Parser<'s> {
...
+                    if !test(c) {
+                        break
+                    } else {
+                        if offset > c.len_utf8() {
+                            // we have some skipping and therefore need to flush
+                            res.push_str(&self.input[sl_start_pos..self.pos]);
+                            res.push(c);
+                            sl_start_pos = self.pos + offset;
+                        }
+                        self.pos += offset;
```
Comment (untitaker) [inline src/vobject/lib.rs]: I suspect that `sl_start_pos` needs to be updated here, even if not `offset > c.len_utf8()`

**Comment by untitaker** [comment]
Also I think you wanted to reference https://github.com/untitaker/rust-vobject/issues/8 instead of #2

File: src/vobject/lib.rs (hunk 3 above)
Comment (gracinet) [inline src/vobject/lib.rs]: No we don't (in three cases), this is me adapting the code as quickly as possible to check if it works overall. It's probably even harmful (wouldn't accomodate cases where CRLF is directly followed by `=` in that specific example)

File: src/vobject/lib.rs (hunk 1 above)
Comment (gracinet) [inline src/vobject/lib.rs]: Yes unecessary, I was just sparing a function call. Actually, I think that `peek_at()` could replace `peek()` entirely in the end.

File: src/vobject/lib.rs (hunk 3 above)
Comment (untitaker) [inline src/vobject/lib.rs]: I suspect you can do `if let Some(('=', _)) = self.peek()`, but unsure.

File: src/vobject/lib.rs (hunk 1 above)
Comment (untitaker) [inline src/vobject/lib.rs]: I'd rather have `peek` around and use `#[inline]` everywhere.

**Comment by untitaker** [comment]
FTR

>for simplicity, I discarded CR completely from peek_at()

That seems reasonable, but please do the simplifications (if you want) in a separate PR

**Comment by gracinet** [comment]
Thanks for the review and the positive appreciation ! I'll take the remarks into account.

Speaking of CR, it wasn't to make the existing code simpler, but rather to avoid make the new one more complicated. I could make the current PR emit CR if not followed by LF in `peek_at` and then a new one that would discards CR again (and also the downstream tests), but the intermediate would be more complicated than the current one (actually tried to do that and gave up after a short while)

As a side note, I also come from a Python background, and funny thing (read your blog post), would like to make a CalDAV server in Rust (partly because I need one, partly because that involves learning a wide array of things).

**Comment by gracinet** [comment]
I did the hiding of `peek_at()` behind `peek()` as an inline, like the idea.
For the various Some('x', 1), this has been abstracted away behind the new `consume_only_char()`

Hope that suits you, will look closer to that CR thing later, have to go right now.

**Comment by untitaker** [comment]
Excellent, thanks!

**Comment by untitaker** [comment]
>As a side note, I also come from a Python background, and funny thing (read your blog post), would like to make a CalDAV server in Rust (partly because I need one, partly because that involves learning a wide array of things). 

"learning a wide array of things" is one side, the other one is "it's incredibly hard".

I recommend SabreDAV as a reference on how to implement things properly, in case the spec is either unclear or incomprehensible.

**Comment by gracinet** [comment]
thanks for the advice ! The devil lies in practical details indeed, I wouldn't have imagined querying to be so complicated. We'll see, I don't mind trashing the project if it becomes really painpful and does not qualify as training anymore.

**Comment by untitaker** [comment]
For a start you can ignore queries and just return all items of the collection, it's correctly handled by most clients (although imperformant).

Feel free to email me questions if you have some.

**Comment by untitaker** [comment]
@gracinet How's it going with your CalDAV project?

**Comment by gracinet** [comment]
@untitaker hey, thanks for asking ! I have a very partial implementation for CardDAV (simpler), but had to put in on standby to get back to more down-to-earth concerns, and now I haven't played with it since around march I think.

If I remember correctly, last time I played with it, I was trying to use a cache to avoid reparsing the collections at each PROPFIND request, which to be fair is not necessary: from the measurements I got, even without such a cache, performance is already an order of magnitude better than radicale's, but it'd be again two orders better with it ;-)

I can show you if you're interested, of course. What I'd really need to get further would probably be a genuine need of it somewhere.

**Comment by untitaker** [comment]
Are you saving every collection as one file each? I don't see a need for
parsing if you mirror the filestructure that is exposed via WebDAV directly
onto the filesystem, i.e. have one file per vcard.

**Comment by gracinet** [comment]
yes, one file per collection is, IIRC what radicale does, but it doesn't matter that much: the clients I've played with (in CardDAV context) would request the properties of all the items at once, or something alike (don't remember that well).

Anyway, it's not very relevant, that was just to say I was looking at something that I knew very well wasn't really needed, one way or the other instead of finishing to implement the standard, because it was an interesting exercise in the language.
